src/core/qgstemporalnavigationobject.cpp

```diff
@@ -312,11 +312,14 @@ QgsTemporalNavigationObject::AnimationState QgsTemporalNavigationObject::animati
   return mPlayBackMode;
 }
 
-long QgsTemporalNavigationObject::findBestFrameNumberForFrameStart( const QDateTime &frameStart ) const
+long long QgsTemporalNavigationObject::findBestFrameNumberForFrameStart( const QDateTime &frameStart ) const
 {
-  long bestFrame = 0;
-  QgsDateTimeRange testFrame = QgsDateTimeRange( frameStart, frameStart ); // creatng an 'instant' Range here
-  for ( long i = 0; i < totalFrameCount(); ++i )
+  long long bestFrame = 0;
+  QgsDateTimeRange testFrame = QgsDateTimeRange( frameStart, frameStart ); // creating an 'instant' Range here
+  // earlier we looped from frame 0 till totalFrameCount() here, but this loop grew gigantic if you switched to for example milliseconds
+  // that is why now we calculate a rough framestart here
+  long long roughFrameStart = std::floor( ( frameStart - mTemporalExtents.begin() ).seconds() / mFrameDuration.seconds() );
```

Comment:
```suggestion
  long long roughFrameStart = std::floor( ( frameStart - mTemporalExtents.begin() ).seconds() / mFrameDuration.seconds() * 30.0 / 31.0 );
```

...with an appropriate explanatory comment, of course!

Reply:
Thanks @nyalldawson

I have been thinking about something like this too, but to have a more precise solution, I think we have to know (take into account) the 'old' frame duration, because THAT determines the 'uncertainty' of the logic. The use of the month here also looks to me as 'fixing one whole, but not closing the water'?

As you said in the earlier comment, the solution should be based on the worse scenario, which (I think) is: going from 'year' (fluid) to 'milliseconds' step. While very theoretical, that is the worst one (and the actual issue we came to this), I think not yet fixed with your proposal...?
Another 'hole' is (maybe?) if we are stepping in units of 2 or 3 (like 3 months...) and moving to 1 second step or so... 

Would be a better solution:
- add the 'old' step duration (be it months or year) as a parameter to the method
- (ony) if years or months (the fluid ones) do some kind of floor preparation as you propose?
- max out the loop to the number of old stepunits which are in the new stepunits...?

Argh, this is getting more and more complex, to me.

I'll try and test your suggestion and see what happens...

Reply:
> going from 'year' (fluid) to 'milliseconds' step.

I don't think so. Because worst case there is we assume a 365.25 day year length, and the actual length is 366 days (ie the frame guess might be 1.00273x too large). But for months its assuming 30 day month when actual length is 31, so frame guess would be 1.0333x too large. 

I think my solution works, in that it will cause the start guess to always be a bit before the actual frame, even in that worst case scenario (because of the floor). But it'll be close enough that the extra iterations won't be noticeable...

Reply:
Ok, I tried/tested (and still testing) with the WMST from:
http://geoservices.knmi.nl/cgi-bin/RADNL_OPER_R___25PCPRR_L3.cgi
But your proposal does not work yet.

I either get: a loop which is (in minutes/seconds) very noticable (like seconds long)
or the loop does not even start, because the roughFrameStart is getting bigger then the calculated totalFrameCount.

I also tried to find your issue by switching between several days on (long OR small) months, and then go to minutes, but the old algorithm actually never failed as the calculated roughFrameStart seems actually always accurate.

Another option is that (as I'm pretty confident that you never do more then a couple of steps), is to maximize the loop to just 10 steps or so.
Then, IF the roughFrameStart is wrong at least it is not freezing the ui, but it will (in your case) go back to step 1. Making the worst case scenario: QGIS behaves like it was before the 'guessing of the old framestep'.

Untill somebody actually finds a real usecase in which the problem you sketch (or at least it is going back to first step while the user is expecting it should stay in the old position)?

Reply:
> because the roughFrameStart is getting bigger then the calculated totalFrameCount

Hmm... that's very odd, because the change should ALWAYS make the frame guess smaller! I can't see how multiplying the result by 0.96774 would ever make it larger....?

Reply:
@nyalldawson you are right, I'm just trying to find your problem. So testing trying etc etc. 
But my most important conclusions:
1) I cannot find a case in which your issue occurs
2) The proposed fix still makes the ui freeze (because 0.96 year is still a lot of (milli)seconds to loop through)
3) I will minimize the loop so it will never do more then 10 steps (as in my tests the logic is always spot on)
3) Else: let's roll back my earlier 'findBestFrameNumberForFrameStart' commit, and go back to always to start...

Another issue that is popping up: the slider is not updated ok: going from years to minutes and being in the middle of the rangeextent, you should stay on the middle of the slider. But you are being moved back and go to the right position after you step one step...  I think it has something to do with the rather complex update logic of the widget...

Reply:
I just realised that we are doing something slightly silly here -- the inconcistent frame sizes will ONLY be an issue if the native units for the frame duration are months, years, decades or centuries. For all the other units your original logic is correct, as there's no fluidity in the actual frame sizes. And if the unit is months or greater, then you're very unlikely to be in the situation that you're looping through millions of frames anyway...! So I think you could revert back to your original logic with an extra check that the interval native type isn't months/..., and we'll be fine to proceed!

Reply:
Thanks @nyalldawson but sorry, which original logic do you mean? 
I do have my original one currently, WITH extra max loop size of 100.
Or do you propose to only do the guessing when unit < month? And else just loop through all? That’s makes sense, I think, just checking...
Second: do you think https://github.com/qgis/QGIS/pull/41421/commits/532028398ae2d725017369e1e1b3079dec3438e3 is ok, or is it a potential ui issue. It fixes the wrong position of the slider if you move to a smaller step unit.

Reply:
>do you think 5320283 is ok, or is it a potential ui issue. It fixes the wrong position of the slider if you move to a smaller step unit.

looks good to me!